Improve Kestrel endpoints doc (#29501)
dotnet/AspNetCore.Docs: aspnetcore/fundamentals/servers/kestrel/endpoints.md

monikerRange: '>= aspnetcore-5.0'
ms.author: riande
ms.custom: mvc
ms.date: 06/14/2023
uid: fundamentals/servers/kestrel/endpoints
---

# Configure endpoints for the ASP.NET Core Kestrel web server

:::moniker range="< aspnetcore-7.0"
[!INCLUDE [](~/includes/not-latest-version.md)]
:::moniker-end

:::moniker range=">= aspnetcore-8.0"

> An endpoint in Kestrel is defined by a combination of address and protocol. The address specifies the network interface and port on which the server listens for incoming requests. This can be a specific IP address, such as localhost or a public IP, along with a port number. The protocol defines the communication standard that is used between the client and server, such as HTTP or HTTPS. Endpoints can be configured to support HTTP or HTTPS or both, depending on the environment (development or production) and the requirements of the application. Endpoints in ASP.NET Core Kestrel provide the infrastructure for listening to incoming requests and routing them to the appropriate middleware.

Reviewer: ```suggestion
Endpoints in ASP.NET Core Kestrel provide the infrastructure for listening to incoming requests and routing them to the appropriate middleware. An endpoint in Kestrel is defined by a combination of address and protocol.

* The address specifies the network interface and port on which the server listens for incoming requests. This can be a specific IP address, such as localhost or a public IP, along with a port number.
* The protocol defines the communication standard that is used between the client and server, such as HTTP or HTTPS, or the HTTP version. Endpoints can be configured to support HTTP or HTTPS or both, depending on the environment (development or production) and the requirements of the application.
```
I think this is a bit better. It describes what an endpoint is used for first, then goes into detail about what an endpoint is.

monikerRange: '>= aspnetcore-5.0'
ms.author: riande
ms.custom: mvc
ms.date: 06/15/2023
uid: fundamentals/servers/kestrel/endpoints
---

# Configure endpoints for the ASP.NET Core Kestrel web server

:::moniker range="< aspnetcore-7.0"
[!INCLUDE [](~/includes/not-latest-version.md)]
:::moniker-end

:::moniker range=">= aspnetcore-8.0"

Endpoints in ASP.NET Core Kestrel provide the infrastructure for listening to incoming requests and routing them to the appropriate middleware. An endpoint in Kestrel is defined by a combination of address and protocol.

> * The address specifies the network interface and port on which the server listens for incoming requests. This can be a specific IP address, such as localhost or a public IP, along with a port number.

Author: ```suggestion
* The address specifies the network interface and port on which the server listens for incoming requests. This can be a specific IP address with or without a port number. If the port isn't specified, a default port is used, typically 80 for HTTP or 443 for HTTPS. The address can also be a host name, such as `localhost`, which resolves to one or more IP addresses.
```

> * The address specifies the network interface and port on which the server listens for incoming requests. This can be a specific IP address with or without a port number. If the port isn't specified, a default port is used, typically 80 for HTTP or 443 for HTTPS. The address can also be a host name, such as `localhost`, which resolves to one or more IP addresses.

Reviewer: ```suggestion
* The address specifies the network interface and port on which the server listens for incoming requests.
```

> * The protocol defines the communication standard that is used between the client and server, such as HTTP or HTTPS, or the HTTP version. Endpoints can be configured to support HTTP or HTTPS or both, depending on the environment (development or production) and the requirements of the application.

Reviewer: No, but seeing this in the docs makes me doubt myself. Who added this?

> * The protocol defines the communication standard that is used between the client and server, such as HTTP or HTTPS, or the HTTP version. Endpoints can be configured to support HTTP or HTTPS or both, depending on the environment (development or production) and the requirements of the application.

Author: We didn't have an introduction for the article, this was my attempt at writing one, with some less than perfect help from AI.

```suggestion
* The protocol defines the communication standard that is used between the client and server, such as HTTP or HTTPS, or the HTTP version. Endpoints can be configured to support HTTP or HTTPS, depending on the environment (development or production) and the requirements of the application.
```

> * The protocol defines the communication standard that is used between the client and server, such as HTTP or HTTPS, or the HTTP version. Endpoints can be configured to support HTTP or HTTPS, depending on the environment (development or production) and the requirements of the application.

Reviewer: ```suggestion
* The protocol defines the communication standard between the client and server, such as HTTP or HTTPS, and the HTTP version. Endpoints can be configured to support HTTP or HTTPS, depending on the environment (development or production) and the requirements of the application.
```

> 

Reviewer: ```suggestion

Kestrel offers many ways to configure endpoints, such as specifying URLs, JSON configuration in `appsettings.json`, or code in application startup.

```

## Default bindings

> ASP.NET Core projects are configured to bind to a random HTTP port between 5000-5300 and a random HTTPS port between 7000-7300. This default configuration is specified in the generated `Properties/launchSettings.json` file and can be overridden. If no ports are specified, Kestrel binds to `http://localhost:5000`.

Reviewer: ```suggestion
ASP.NET Core projects are configured to bind to a random HTTP port between 5000-5300 and a random HTTPS port between 7000-7300. This default configuration is specified in the generated `Properties/launchSettings.json` file and can be overridden. The `launchSetting.json` file is only used in local development.

If there is no endpoint configuration then Kestrel binds to `http://localhost:5000`.
```

## Configure endpoints with URLs

The following sections explain how to configure endpoints using the:

* `ASPNETCORE_URLS` environment variable.
* `--urls` command-line argument.
* `urls` host configuration key.
* <xref:Microsoft.AspNetCore.Hosting.HostingAbstractionsWebHostBuilderExtensions.UseUrls%2A> extension method.
* <xref:Microsoft.AspNetCore.Builder.WebApplication.Urls?displayProperty=nameWithType> property.

### URL formats

> The URLs indicate the IP addresses or host addresses with ports and protocols that the server should listen on. URLs can be in any of the following formats.

Author: ```suggestion
The URLs indicate the IP addresses or host addresses with ports and protocols that the server should listen on. The port can be omitted if it's the default for the protocol (typically 80 and 443). URLs can be in any of the following formats.
```

* IPv4 address with port number

  ```
  http://65.55.39.10:80/
  ```

  `0.0.0.0` is a special case that binds to all IPv4 addresses.

* IPv6 address with port number

  ```
  http://[0:0:0:0:0:ffff:4137:270a]:80/
  ```

  `[::]` is the IPv6 equivalent of IPv4 `0.0.0.0`.

* Host name with port number

  ```
  http://contoso.com:80/
  http://*:80/
  ```

>   Host names, `*`, and `+`, aren't special. Anything not recognized as a valid IP address or `localhost` binds to all IPv4 and IPv6 IP addresses. To bind different host names to different ASP.NET Core apps on the same port, use [HTTP.sys](xref:fundamentals/servers/httpsys) or a reverse proxy server.

Author: ```suggestion
  Anything not recognized as a valid IP address or `localhost` is treated as a wildcard that binds to all IPv4 and IPv6 addresses. Some people like to use `*` or `+` to be more explicit. To bind different host names to different ASP.NET Core apps on the same port, use [HTTP.sys](xref:fundamentals/servers/httpsys) or a reverse proxy server.
```

Multiple URL prefixes can be specified by using a semicolon (`;`) delimiter:

```
http://*:5000;http://localhost:5001;https://hostname:5002
```

For more information, see [Override configuration](xref:fundamentals/host/web-host#override-configuration).

### Port 0

When the port number `0` is specified, Kestrel dynamically binds to an available port. The following example shows how to determine which port Kestrel bound at runtime:

:::code language="csharp" source="~/fundamentals/servers/kestrel/samples/6.x/KestrelSample/Snippets/Program.cs" id="snippet_IServerAddressesFeature":::

Dynamically binding a port isn't available in some situations:

* <xref:Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerOptions.ListenLocalhost%2A?displayProperty=nameWithType>
* Binding TCP-based HTTP/1.1 or HTTP/2, and QUIC-based HTTP/3 together.

### Specify ports only

[!INCLUDE [http-ports](~/includes/http-ports.md)]

> ### HTTPS URL prefixes

Reviewer 2: Is this standard terminology?  I'm not familiar with it, but I think it's referring to the "https://" at the beginning of the configured address?

Author: It refers to the whole URL string used to define the end point.  It's been used that way as long as we've had this Kestrel documentation and is also used that way in the [HTTP.sys](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/servers/httpsys?view=aspnetcore-7.0#configure-windows-server) and [Web host](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/host/web-host?view=aspnetcore-7.0#server-urls) articles. Its use here may have originated with its use in the HTTP.sys API. If it's causing confusion maybe we should replace "URL prefix" with "URL" throughout the doc.

> HTTPS URL prefixes can be used to define endpoints only if a default certificate is provided in the HTTPS endpoint configuration. For example, use <xref:Microsoft.AspNetCore.Server.Kestrel.KestrelServerOptions> configuration or a configuration file, as shown [later in this article](#configure-certificates-in-appsettingsjson).

Reviewer 2: Is that true?  I think you can pick up a default cert from the cert store?

Author: What should we say here, then?

* In the Nginx configuration file, set the `server` > `location` > `proxy_pass` entry to `http://unix:/tmp/{KESTREL SOCKET}:/;`. `{KESTREL SOCKET}` is the name of the socket provided to <xref:Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerOptions.ListenUnixSocket%2A> (for example, `kestrel-test.sock` in the preceding example).
* Ensure that the socket is writeable by Nginx (for example, `chmod go+w /tmp/kestrel-test.sock`).

### Dynamic port binding

For information about dynamic port binding, see [Port 0](#port-0) earlier in this article.

### Connection logging

Call <xref:Microsoft.AspNetCore.Hosting.ListenOptionsConnectionLoggingExtensions.UseConnectionLogging%2A> to emit Debug level logs for byte-level communication on a connection. Connection logging is helpful for troubleshooting problems in low-level communication, such as during TLS encryption and behind proxies. If `UseConnectionLogging` is placed before `UseHttps`, encrypted traffic is logged. If `UseConnectionLogging` is placed after `UseHttps`, decrypted traffic is logged. This is built-in [Connection middleware](#connection-middleware).

> :::code language="csharp" source="~/fundamentals/servers/kestrel/samples/6.x/KestrelSample/Snippets/Program.cs" id="snippet_ConfigureKestrelUseConnectionLogging":::

Reviewer: I think this section should move to be a child of the "Connection middleware" section

## Configure HTTPS

> If [URL prefixes](#configure-endpoints-with-urls) are used to define endpoints, HTTPS can be used only if a default certificate is provided in HTTPS endpoint configuration. For example, use <xref:Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerOptions> configuration or a configuration file as shown [earlier in this article](#configure-certificates-in-appsettingsjson).

Reviewer: This information is specific to one way of configuring endpoints (via URLs).

A more general introduction to configuring HTTPS with Kestrel is needed here. At the very least some content that says:

* Kestrel supports securing endpoints with HTTPS.
* The way you configure HTTPS depends on how endpoints are defined.

I'll need to think about this area more.

Author: Would something like this make sense:

> ... The way you configure HTTPS depends on how endpoints are defined:
> * If by using URLs, provide a default certificate
> * If by using configuration (e.g. appsettings.json), see the sections on using configuration for HTTPS endpoints:
>   * [Links to sections on using config]
> * If by using code (Listen API), see the sections on using code for HTTPS endpoints
>   * [Links to sections on using code]

To create a development certificate, use the [dev-certs tool](/dotnet/core/tools/dotnet-dev-certs). The tool is automatically installed when the [.NET SDK](/dotnet/core/sdk) is installed.

The development certificate is available only for the user that generates the certificate. Some browsers require granting explicit permission to trust the local development certificate.

For information about configuring HTTPS, such as specifying SSL certificates and SSL/TLS protocols, see the following sections of this article:

* [Configure certificates in appsettings.json](#configure-certificates-in-appsettingsjson)
* [Configure client certificates in appsettings.json](#configure-client-certificates-in-appsettingsjson)
* [Certificate sources](#certificate-sources)
* [Configure HTTPS defaults in code](#configure-https-defaults-in-code)
* [Configure certificates in code](#configure-certificates-in-code)
* [Configure client certificates in code](#configure-client-certificates-in-code)
* [Configure endpoints using Server Name Indication (SNI)](#configure-endpoints-using-server-name-indication)

> * [Connection middleware)](#connection-middleware)

Reviewer: ```suggestion
* [Connection middleware](#connection-middleware)
```

}
```

> ### Configure client certificates in appsettings.json

Reviewer 2: This feels like a new topic.  I'd probably put it after all the ways to configure the server certificate.

Reviewer 2: Probably together with "Configure client certificates in code"

Author: The proposed plan in #21117 was to group all the using-config sections together and the using-code sections together.

Reviewer 2: Ping @JamesNK

Reviewer: I don't think there is a perfect answer.

We have different ways to configure: URLs, config, code
We have different things to configure: address, HTTPS, protocol

We have to pick one.

:::code language="csharp" source="~/fundamentals/servers/kestrel/samples/6.x/KestrelSample/Snippets/Program.cs" id="snippet_ConfigureEndpointDefaultsConfigureHttpsDefaults":::

#### Configure endpoint defaults in code

[`ConfigureEndpointDefaults(Action<ListenOptions>)`](xref:Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerOptions.ConfigureEndpointDefaults(System.Action{Microsoft.AspNetCore.Server.Kestrel.Core.ListenOptions}) Specifies a configuration `Action` to run for each specified endpoint. Calling `ConfigureEndpointDefaults` multiple times replaces prior `Action` instances with the last `Action` specified.

:::code language="csharp" source="~/fundamentals/servers/kestrel/samples/6.x/KestrelSample/Snippets/Program.cs" id="snippet_ConfigureEndpointDefaults":::

> [!NOTE]
> Endpoints created by calling <xref:Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerOptions.Listen%2A> **before** calling <xref:Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerOptions.ConfigureEndpointDefaults%2A> won't have the defaults applied.

> #### Configure HTTPS defaults in code

Reviewer 2: This feels weird, coming after extensive discussion of how to set up certificates.

Author: That is a result of the plan in #21117 to separate using-config and using-code sections. This section is an h4 under change defaults using code h3, which is under configure endpoints in code h2.

Author: The thought behind this structure is that people won't typically read through this doc sequentially from beginning to end -- they'll go directly to the parts that apply to them and skip the rest.

* [Configure client certificates in appsettings.json](#configure-client-certificates-in-appsettingsjson)
* [Certificate sources](#certificate-sources)
* [Configure HTTPS defaults in code](#configure-https-defaults-in-code)
* [Configure certificates in code](#configure-certificates-in-code)
* [Configure client certificates in code](#configure-client-certificates-in-code)
* [Configure endpoints using Server Name Indication (SNI)](#configure-endpoints-using-server-name-indication)
* [Connection middleware](#connection-middleware)
* [Configure endpoint protocols](#configure-endpoint-protocols)

To create a development certificate, use the [dev-certs tool](/dotnet/core/tools/dotnet-dev-certs). The tool is automatically installed when the [.NET SDK](/dotnet/core/sdk) is installed.

The development certificate is available only for the user that generates the certificate. Some browsers require granting explicit permission to trust the local development certificate.

Many ASP.NET Core project templates configure apps to run on HTTPS by default and include [HTTPS redirection and HSTS support](xref:security/enforcing-ssl).

## Configure endpoints using Server Name Indication

> [Server Name Indication (SNI)](https://tools.ietf.org/html/rfc6066#section-3) can be used to host multiple domains on the same IP address and port. For SNI to function, the client sends the host name for the secure session to the server during the TLS handshake so that the server can provide the correct certificate. The client uses the furnished certificate for encrypted communication with the server during the secure session that follows the TLS handshake.

Author: ```suggestion
[Server Name Indication (SNI)](https://tools.ietf.org/html/rfc6066#section-3) can be used to host multiple domains on the same IP address and port. You might want to do this to conserve resources by serving multiple sites from one server.

For SNI to function, the client sends the host name for the secure session to the server during the TLS handshake so that the server can provide the correct certificate. The client uses the furnished certificate for encrypted communication with the server during the secure session that follows the TLS handshake.
```

* The configuration section for each endpoint is available on the options in the <xref:Microsoft.AspNetCore.Server.Kestrel.KestrelConfigurationLoader.Endpoint%2A> method so that custom settings may be read.
* Multiple configurations can be loaded by calling <xref:Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerOptions.Configure(Microsoft.Extensions.Configuration.IConfiguration)?displayProperty=nameWithType> again with another section. Only the last configuration is used, unless `Load` is explicitly called on prior instances. The metapackage doesn't call `Load` so that its default configuration section may be replaced.
* `KestrelConfigurationLoader` mirrors the `Listen` family of APIs from `KestrelServerOptions` as `Endpoint` overloads, so code and config endpoints can be configured in the same place. These overloads don't use names and only consume default settings from configuration.

## Configure endpoints in code

<xref:Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerOptions> provides methods for configuring endpoints:

* <xref:Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerOptions.Listen%2A>

> * <xref:Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerOptions.ListenUnixSocket%2A>

Reviewer: ```suggestion
* <xref:Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerOptions.ListenUnixSocket%2A>
* <xref:Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerOptions.ListenNamedPipe%2A>
```

:::code language="csharp" source="~/fundamentals/servers/kestrel/samples/6.x/KestrelSample/Snippets/Program.cs" id="snippet_Listen":::

The example configures HTTPS for an endpoint with <xref:Microsoft.AspNetCore.Server.Kestrel.Core.ListenOptions>. Use the same API to configure other Kestrel settings for specific endpoints.

[!INCLUDE [How to make an X.509 cert](~/includes/make-x509-cert.md)]

### Bind to a Unix socket

Listen on a Unix socket with <xref:Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerOptions.ListenUnixSocket%2A> for improved performance with Nginx, as shown in this example:

:::code language="csharp" source="~/fundamentals/servers/kestrel/samples/6.x/KestrelSample/Snippets/Program.cs" id="snippet_ListenUnixSocket":::

* In the Nginx configuration file, set the `server` > `location` > `proxy_pass` entry to `http://unix:/tmp/{KESTREL SOCKET}:/;`. `{KESTREL SOCKET}` is the name of the socket provided to <xref:Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerOptions.ListenUnixSocket%2A> (for example, `kestrel-test.sock` in the preceding example).

> * Ensure that the socket is writeable by Nginx (for example, `chmod go+w /tmp/kestrel-test.sock`).

Reviewer: Move these sections to under neither "Configure endpoints in code"

Author: Done

#### SNI with `ServerOptionsSelectionCallback`

Kestrel supports additional dynamic TLS configuration via the `ServerOptionsSelectionCallback` callback. The callback is invoked once per connection to allow the app to inspect the host name and select the appropriate certificate and TLS configuration. Default certificates and `ConfigureHttpsDefaults` aren't used with this callback.

:::code language="csharp" source="~/fundamentals/servers/kestrel/samples/6.x/KestrelSample/Snippets/Program.cs" id="snippet_ServerOptionsSelectionCallback":::

#### SNI with `TlsHandshakeCallbackOptions`

Kestrel supports additional dynamic TLS configuration via the `TlsHandshakeCallbackOptions.OnConnection` callback. The callback is invoked once per connection to allow the app to inspect the host name and select the appropriate certificate, TLS configuration, and other server options. Default certificates and `ConfigureHttpsDefaults` aren't used with this callback.

:::code language="csharp" source="~/fundamentals/servers/kestrel/samples/6.x/KestrelSample/Snippets/Program.cs" id="snippet_TlsHandshakeCallbackOptions":::

> ## IIS endpoint configuration

Reviewer 2: Why is this in an article about Kestrel configuration?

Reviewer: +1. I'd like to understand how it ended up here. If there isn't a good reason then I say dump it.

Author: I don't know when it was added.  I'll dump it.

:::code language="csharp" source="~/fundamentals/servers/kestrel/samples/6.x/KestrelSample/Snippets/Program.cs" id="snippet_ServerOptionsSelectionCallback":::

#### SNI with `TlsHandshakeCallbackOptions`

Kestrel supports additional dynamic TLS configuration via the `TlsHandshakeCallbackOptions.OnConnection` callback. The callback is invoked once per connection to allow the app to inspect the host name and select the appropriate certificate, TLS configuration, and other server options. Default certificates and `ConfigureHttpsDefaults` aren't used with this callback.

:::code language="csharp" source="~/fundamentals/servers/kestrel/samples/6.x/KestrelSample/Snippets/Program.cs" id="snippet_TlsHandshakeCallbackOptions":::

## IIS endpoint configuration

When using IIS, the URL bindings for IIS override bindings are set by either `Listen` or `UseUrls`. For more information, see [ASP.NET Core Module](xref:host-and-deploy/aspnet-core-module).

## Connection middleware

> Custom connection middleware can filter TLS handshakes on a per-connection basis for specific ciphers if necessary.
> 
> The following example throws <xref:System.NotSupportedException> for any cipher algorithm that the app doesn't support. Alternatively, define and compare <xref:Microsoft.AspNetCore.Connections.Features.ITlsHandshakeFeature.CipherAlgorithm%2A?displayProperty=nameWithType> to a list of acceptable cipher suites.
> 
> No encryption is used with a <xref:System.Security.Authentication.CipherAlgorithmType.Null?displayProperty=nameWithType> cipher algorithm.
> 
> :::code language="csharp" source="~/fundamentals/servers/kestrel/samples/6.x/KestrelSample/Snippets/Program.cs" id="snippet_ConfigureKestrelMiddleware":::

Reviewer: The description here should describe what connection middleware is. Sort of like the description at https://learn.microsoft.com/en-us/aspnet/core/fundamentals/middleware/?view=aspnetcore-7.0 for request middleware.

Something like:

```suggestion
Kestrel endpoints support connection middleware. Connection middleware is software that is assembled into a connection pipeline and runs when Kestrel receives a new connection. Each component:

* Chooses whether to pass the request to the next component in the pipeline.
* Can perform work before and after the next component in the pipeline.

Connection delegates are used to build the connection pipeline. Connection delegates are configured with the https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.server.kestrel.core.listenoptions.use method.

Connection middleware is similar to [ASP.NET Core request middleware](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/middleware/?view=aspnetcore-7.0). The difference is the connection middleware runs per-connection instead of per-request.

### Create custom connection middleware

The following example shows a custom connection middleware that can filter TLS handshakes on a per-connection basis for specific ciphers if necessary. The middleware throws <xref:System.NotSupportedException> for any cipher algorithm that the app doesn't support. Alternatively, define and compare <xref:Microsoft.AspNetCore.Connections.Features.ITlsHandshakeFeature.CipherAlgorithm%2A?displayProperty=nameWithType> to a list of acceptable cipher suites.

No encryption is used with a <xref:System.Security.Authentication.CipherAlgorithmType.Null?displayProperty=nameWithType> cipher algorithm.

:::code language="csharp" source="~/fundamentals/servers/kestrel/samples/6.x/KestrelSample/Snippets/Program.cs" id="snippet_ConfigureKestrelMiddleware":::
```

The links need to be fixed up. Make other changes and clean up as necessary.

Reviewer: Connection middleware is configured against an endpoint, but there is an option to split it out into a new `Connection middleware` page under Kestrel. That would reduce how much content is on the endpoint page.

Moving connection middleware to a new page could be a follow-up PR.
